### PR TITLE
Feature/export aspect

### PR DESCRIPTION
**Pull request by gmrukwa**
This PR introduces *export aspect*. It allows to export the transformed dataset under any name specified by the user for any further analysis.

It is backward compatible in a way, that regenerates text representation of transformed data, when missing after analysis in previous version.

DO NOT MERGE BEFORE #3.

image

### CODE REVIEW

**Comment by gmrukwa** [comment]
[t-SNE export aspect](https://trello.com/c/wnVgewxm/466-t-sne-export-aspect)

**Review by Daroslaw — DISMISSED** [review]

File: common.py [hunk 1]
```diff
@@ -14,7 +14,59 @@
...
+    route = path.split('.')
+    try:
+        for way in route:
+            variables = variables[way]
+    except KeyError as ex:
+        raise ValueError('Unavailable variable: %s' % path) from ex
+    return variables
+
+
+def with_open(mode='r', buffering=None, encoding=None, errors=None, newline=None, closefd=True):
```
Comment (mg6) [inline common.py]: Missing `*args, **kwargs` params to inner [`open`](https://docs.python.org/3/library/functions.html#open); you omitted `opener` here. You should account for possible future extension of stdlib functions this way.

Comment (mg6) [reply common.py]: More generally, I think this function introduces unnecessary complexity around simple `with open(path) as f` construct, which could be used for `dumps_txt` function, that can be as simple as:
```py
def dumps_txt(path, data: ty.Dataset, *args, **kwargs):
    with open(path, 'w', *args, **kwargs) as f:
        return dump_txt(f, data)
```

File: data_utils.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def push_to_repo(path: str, name: str):
+    """Push text dataset to data store"""
+    extension = os.path.splitext(path)[1].lower()
+    if extension != '.txt':
+        raise NotImplementedError('Only .txt are supported now. Was: %s' %
+                                  extension)
+    dst_dir = os.path.join(cmn.DATA_ROOT, name, 'text_data')
+    os.makedirs(dst_dir)
```
Comment (mg6) [inline data_utils.py]: Should we account for already existing directory here? This function will raise in current configuration.

**Review by mg6 — DISMISSED** [review]

File: aspect/_export.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+def export(analysis_id: str) -> common.Response:
+    """Export transformed dataset to a common data store
+
+    Arguments:
+        analysis_id - (str) ID of transform to use
+
+    Implicit arguments (should be provided in JSON request):
+        target_name - (str) name of the dataset created in common data store
+    """
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply aspect/_export.py]: Sure, I'll refine it. However these implicit arguments are nowhere specified in Google docstring convention, while I think it is crucial to write about them.

File: aspect/_export.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+def ensure_dataset(root: str) -> str:
+    """Return a path to an existing dataset after transform"""
+    dataset_path = os.path.join(root, 'data.txt')
+    if not os.path.exists(dataset_path):
+        regenerate_dataset(dataset_path, root)
+    return dataset_path
+
+
+EMPTY_TABLE = repr({"columns": [], "data": []}).replace('\'', '"'), 200
```
[the remark this answers is not in the document]
Comment (gmrukwa) [reply aspect/_export.py]: I'd like to, but unfortunately Flask isn't smart enough to handle this. But the clue hits the point: I should think of separation of these concerns a little bit better.

File: aspect/_export.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mg6) [reply aspect/_export.py]: Not referring to implicit args at all here, they are useful and should be preserved.

Pay attention to Google-style *type* notation instead. :-)

File: aspect/_export.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mg6) [reply aspect/_export.py]: It is, absolutely; see [this](https://flask-restful.readthedocs.io/en/latest/quickstart.html) and [this](http://flask-json.readthedocs.io/en/latest/#basic-usage).

File: common.py (hunk 1 above)
Comment (mg6) [reply common.py]: Unnecessary complexity in my opinion, since you save 2 or 3 lines only. Not worth it.

Comment (mg6) [reply common.py]: And it has already bitten you, as shown by 781537c.

File: data_utils.py (hunk 2 above)
Comment (mg6) [reply data_utils.py]: Okay, should be fine as is.

File: common.py (hunk 1 above)
Comment (gmrukwa) [reply common.py]: It has bitten - that's true. But on the other hand, DRY? We will spam ourselves with functions consisting only of

```
def dumps_something(path, data, *args, **kwargs):
    with open(path, 'w', *args, **kwargs) as f:
        return dump_something(f, data)
```

File: aspect/_export.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gmrukwa) [reply aspect/_export.py]: These are additional dependencies we will have to introduce, to just support JSON output this way... I'll dig a bit.

**Review by mg6 — APPROVED** [review]

**Review by Daroslaw — APPROVED** [review]

**Pull request merged**